Upgrade django-two-factor-auth from 1.13.2 to 1.16.0

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
I updated the text on the 2FA profile page to better reflect the remove and reset options.
![Screenshot from 2023-12-27 17-39-28](https://github.com/dimagi/commcare-hq/assets/15785053/e9a26edd-a39c-40ee-9c75-1cab89be3957)
## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/SAAS-15067

[django-two-factor-auth changelog](https://github.com/jazzband/django-two-factor-auth/blob/master/CHANGELOG.md)

This is a fairly large change to our 2FA setup and login workflows which will require detailed QA.

This is my second attempt at a PR for upgrading to v1.15.5 (first attempt [here](https://github.com/dimagi/commcare-hq/pull/33700)). As noted there, a lot of unforeseen issues came up when performing this upgrade that started to lead me down a path to a messy commit history.

The change/commit that is most central to this PR and that I'm most eager to get input on is eac8eb347e0fb249e4d8bc741872121f73cdb33a. `django-two-factor-auth` changed the way custom forms are specified for 2FA workflows. Rather than specify the custom form in a `form_list` attribute on the custom view class, we need to use the `registry` object that lives within the library to register custom 2FA methods (which reference our custom forms). Notably, the `form_list` attribute still exists for non-dynamic forms/methods. ~The logic I have in the `setup_two_factor` decorator basically ensures our custom methods are registered every time a new view is created by a request.~

Review by commit 🐟 I did my best to break this up into more digestible commits, but it was a bit of a challenge so apologies if it is hard to review.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
I've tested the various 2FA workflows locally and on staging, specifically:
- 2FA setup for generator, call, and sms
- login with 2FA enabled (and disabled)
- setup backup phone
- reset 2FA
- remove 2FA

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Thanks to Matt!
- corehq.apps.settings.tests.test_views:TwoFactorSetupView_RegisteredMethods_Tests
- corehq.apps.settings.tests.test_views:TwoFactorPhoneSetupViewTests
- corehq.apps.settings.tests.test_views:TwoFactorProfileView_Context_Tests
- corehq.apps.settings.tests.test_forms:TestHQTwoFactorMethodForm

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
https://dimagi-dev.atlassian.net/browse/QA-5782

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [x] The migrations in this code can be safely applied first independently of the code
I tested this locally and believe this is the case (django-two-factor-auth squashed and moved existing migrations to a new app, but that is all).
<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->
You will also need to rollback the following PRs:
-  https://github.com/dimagi/commcare-hq/pull/33925
- https://github.com/dimagi/commcare-hq/pull/34292

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Upgrade django-two-factor-auth to 1.15.5

- run 'make requirements'
- run './manage.py makemigrations --lock-update'

gherceg committed:
Resolve breaking changes in v1.14.0

- add two_factor.plugins.phonenumber to INSTALLED_APPS
- phone specific logic was moved to two_factor.plugins.phonenumber
- the two_factor template tag was renamed to phonenumber
- fix failing tests due to new models

See https://github.com/jazzband/django-two-factor-auth/blob/master/CHANGELOG.md#changed-3

gherceg committed:
Fix TwoFactorPhoneSetupView

- two_factor:phone_create and phone_delete urls were updated in 1.15.5
  (removed backup)
- PhoneSetupView changed key 'method' to 'setup'

gherceg committed:
Fix HQLoginView and related views

- use setup_two_factor decorator to register custom methods

gherceg committed:
Create custom HQ two factor methods

gherceg committed:
Fix TwoFactorSetupView and related views

- copy updated setup template from django-two-factor-auth to get up to
  date steps along with minor UI improvements
- pass key and user parameters explicitly to HQTOTPDeviceForm super
  class
- pass device explicitly to HQDeviceValidationForm
- override default validation form with custom form in get_form_list
- clear cached attribute on user after deleting device (TwoFactorResetView)
- update tests to conform to new method registration instead of form
  kwargs

gherceg committed:
Update wording for TwoFactorProfileView

gherceg committed:
Fix lint errors

- 'django.urls.get_resolver' imported but unused (F401)

millerdev commented on migrations.lock:
```diff
@@ -1094,13 +1096,7 @@ translations
- 0005_auto_20160224_0450
- 0006_phonedevice_key_default
- 0007_auto_20201201_1019
+ (no migrations)
```
This is a bit surprising. Do you know why these went away? Were they squashed and moved to the `phonenumber` app?

gherceg replied:
Yeah my understanding is they were squashed and moved to the `phonenumber` app to prevent new projects from needing to include the now optional `phonenumber_field` dependency when using this library.

millerdev commented on corehq/apps/hqwebapp/decorators.py:
```diff
@@ -236,6 +237,22 @@ def _inner(request, *args, **kwargs):
+    Use this decorator on the dispatch method of a TemplateView subclass
+
+    We use custom forms for two factor setup and authentication
+    This decorate should be use on any view that depends on two factor
```
```suggestion
    This decorator should be use on any view that depends on two factor
```

gherceg replied:
No longer relevant as this decorator was removed in 4ccbb9f9c74ee16ae5179cd22299b53e1dbdcabf

millerdev commented on corehq/apps/hqwebapp/utils/two_factor.py:
```diff
@@ -0,0 +1,34 @@
+
+
+def user_can_use_phone(user):
+    if not settings.ALLOW_PHONE_AS_DEFAULT_TWO_FACTOR_DEVICE:
```
I think we had been trying to get away from phone as two-factor device, although I see we still set this to `True` on prod. Is it possible we do not use this functionality anymore? (Probably check with product.)

This question may be totally misguided. Maybe phone-two-factor is fine on domains that pay for or have their own SMS gateways?

gherceg replied:
Yeah I'm pretty sure we are not in a position right now to remove support for phone methods and that that would be outside of the scope of this work, but I'll ask anyway to get a pulse check on how 2FA is used across environments/projects. I think the push was specifically for defaulting to generator apps and requiring projects to seek out using less secure phone methods, but not forbidding it.

gherceg replied:
I asked around about this and Daniel and I talked about it briefly, but it seems to make the most sense to leave this in for now. FWIW, we do seem to be closer to being able to remove support for 2FA phone methods than I thought.

millerdev commented on corehq/apps/hqwebapp/decorators.py:
```diff
@@ -236,6 +237,22 @@ def _inner(request, *args, **kwargs):
+    @wraps(view_func)
+    def _inner(view, request, *args, **kwargs):
+        user = request.couch_user if hasattr(request, 'couch_user') else None
+        register_two_factor_methods(user)
```
Why do these need to be registered on every request? (Nothing I can see in the registration code appears to reference the view.) Could they be registered once on Django startup, maybe in an app ready handler?

gherceg replied:
Yeah this may have been an oversight on my part. Related to your next comment about removing the allow_phone_2fa logic, at one point I had the impression that I had to fully rely on registering/unregistering methods, but it seems quite possible and ideal that we could register this once and keep our logic to determine which methods are available to the user. I'll look into this.

gherceg replied:
Fixed in 4ccbb9f9c74ee16ae5179cd22299b53e1dbdcabf

millerdev commented on corehq/apps/settings/forms.py:
```diff
@@ -142,16 +141,8 @@ def clean_token(self):
-    def __init__(self, *, allow_phone_2fa, **kwargs):
-        super().__init__(**kwargs)
-        if not allow_phone_2fa:
-            # Block people from setting up the phone method as their default
```
Was this block moved somewhere else? From my naive perspective it appears to have been removed, which doesn't seem right.

gherceg replied:
See 4ccbb9f9c74ee16ae5179cd22299b53e1dbdcabf

gherceg committed:
Register custom 2FA methods on Django app startup

Replaces the need for the setup_two_factor decorator as the custom
methods are setup at the app level, and the HQTwoFactorMethodForm
controls which methods are available to the user.

The one side effect is that the user encounters the
HQTwoFactorMethodForm regardless of if they are allowed to use phone
methods for 2FA, whereas previously the two factor setup wizard was
smart enough to skip the method step if only one method was registered.

gherceg commented:
@millerdev new and improved iteration on 4ccbb9f9c74ee16ae5179cd22299b53e1dbdcabf. It's a bit larger than I had hoped, but adding back in the `allow_phone_2fa` went hand in hand with the change to register the custom methods on startup only.

I added tests to verify that the custom methods are registered when `register_two_factor_methods` is called in corehq.apps.hqwebapp.tests.test_apps:RegisterCustom2FAMethodsTests. Since that code runs during django startup, I couldn't mock the settings referenced at that point for each test, so the test only ensures that `register_two_factor_methods` behaves as expected.

millerdev commented on corehq/apps/hqwebapp/apps.py:
```diff
@@ -7,3 +8,35 @@ class HqWebAppConfig(AppConfig):
+    registry.register(HQGeneratorMethod())
+
+    # default phone methods are registered when django starts up
+    # https://github.com/jazzband/django-two-factor-auth/blob/master/two_factor/plugins/phonenumber/apps.py#L19-L30
```
```suggestion
    # https://github.com/jazzband/django-two-factor-auth/blob/1.15.5/two_factor/plugins/phonenumber/apps.py#L19-L30
```

millerdev commented on corehq/apps/settings/views.py:
```diff
@@ -402,24 +402,25 @@ class TwoFactorSetupView(BaseMyAccountView, SetupView):
+    def get_form_kwargs(self, step=None):
+        kwargs = super().get_form_kwargs(step)
+        if step == 'method':
+            kwargs.setdefault('allow_phone_2fa', user_can_use_phone(self.request.couch_user))
```
The use of `setdefault` makes it appear that there may be some scenario where _phone 2FA is allowed_ when the _user cannot use phone_. Can you explain that scenario? Otherwise, maybe this should do something like
```suggestion
            kwargs['allow_phone_2fa'] = user_can_use_phone(self.request.couch_user)
```

gherceg replied:
Hmm good point. I was just reverting to what this was previously without much thought. Looking deeper, it looks like @mjriley initially added this with the idea of passing a 'user' object in ([here](https://github.com/dimagi/commcare-hq/commit/5f2523a80226#diff-3240f7d1a05a80cb57e5cd13140f90b74330085c1326cccbbb34733f19f9efaeR378)) but ultimately ended up passing in 'allow_phone_2fa' on [that PR](https://github.com/dimagi/commcare-hq/pull/30347/files#diff-3240f7d1a05a80cb57e5cd13140f90b74330085c1326cccbbb34733f19f9efaeR397-R400). There are other scenarios where having an existing backup phone overrides the result of `user_can_use_phone` ([here](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/settings/views.py#L395-L398)), but I think in the context of this form which is specifically used when setting up 2FA for the first time or resetting it, we want to strictly prevent users who should not be using phones from doing so. I'll give Matt a chance to respond in case there is some other context I'm missing, but otherwise can make the change you've suggested to clear this up.

mjriley replied:
It looks like you've done good digging. I don't remember how the forms worked together, but yeah, the intent was to shut off phone access unless they already had backup phones set. Daniel's approach sounds good here

gherceg replied:
Noting that this was resolved in d94b897253b993dbba6cd400627413b6f79daf6b

millerdev reviewed (commented):
> the user encounters the `HQTwoFactorMethodForm` regardless of if they are allowed to use phone methods for 2FA

Is this a user-facing change, or just a technical detail of the implementation that does not change behavior?

gherceg committed:
Reference 1.15.5 release in github link

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

gherceg commented:
> > the user encounters the HQTwoFactorMethodForm regardless of if they are allowed to use phone methods for 2FA

> Is this a user-facing change, or just a technical detail of the implementation that does not change behavior?

This is referring to what could have been, so actually current behavior is preserved now. I should have made that more clear.

mjriley commented on corehq/apps/settings/tests/test_forms.py:
```diff
@@ -1,32 +1,54 @@
+
     def test_when_phone_support_is_enabled_all_options_are_shown(self):
-        form = HQTwoFactorMethodForm(allow_phone_2fa=True)
+        self.mock_get_methods.return_value = [HQGeneratorMethod(), HQPhoneCallMethod(), HQSMSMethod()]
```
Would it be possible to access the registry directly in this test, rather than mock out the method return values? Mocking out the values seems to change the meaning of this test. Prior, it verified that the form with the given constructor arguments would show the expected options. With the options mocked, it doesn't feel safe to make that guarantee anymore

gherceg replied:
Apologies for the late response but given the changes I've made this no longer applies and the test has been reverted to its previous state. Thanks for flagging.

mjriley commented on corehq/apps/settings/tests/test_views.py:
```diff
@@ -105,29 +107,39 @@ def _create_view_for_user(self, user):
-        view = self._create_view_for_user(user)
-        self.assertFalse(view.get_form_kwargs(step='method')['allow_phone_2fa'])
+        self._create_view_for_user(user)
+        self.assertCountEqual(['generator'], [m.code for m in registry.get_methods()])
```
Would it make sense to assert that 'sms' and 'call' were not in the registered methods, rather than 'generator' was? If 'generator' stopped being part of the results from the API, I don't think that should break this test, right?

gherceg replied:
Great point! I ended up reverting this anyway once I started registered methods in app startup instead of the view decorator, but that is great feedback.

mjriley commented on corehq/apps/settings/tests/test_views.py:
```diff
@@ -105,29 +107,39 @@ def _create_view_for_user(self, user):
-        view = self._create_view_for_user(user)
-        self.assertTrue(view.get_form_kwargs(step='method')['allow_phone_2fa'])
+        self._create_view_for_user(user)
+        self.assertCountEqual(['generator', 'sms', 'call'], [m.code for m in registry.get_methods()])
```
Similar to the above, should this change to only assert that 'sms' and 'call' are in the returned methods?

gherceg replied:
Same as above, reverted this change in 4ccbb9f9c74e

mjriley commented on corehq/apps/settings/tests/test_views.py:
```diff
@@ -105,29 +107,39 @@ def _create_view_for_user(self, user):
+        self._create_view_for_user(user)
+        self.assertCountEqual(['generator'], [m.code for m in registry.get_methods()])
+
+    def test_custom_methods_are_registsered(self):
```
Nit: typo with `registsered`

gherceg replied:
Reverted in 4ccbb9f9c74e

mjriley commented on corehq/apps/settings/tests/test_views.py:
```diff
@@ -105,29 +107,39 @@ def _create_view_for_user(self, user):
+        self.assertEqual(HQPhoneCallMethod, type(call_method))
+
+        sms_method = registry.get_method('sms')
+        self.assertEqual(HQSMSMethod, type(sms_method))
```
Nit: I think our assertions are generally of the `self.assert(actual, expected)` form. Flipping them to `self.assertEqual(type(generator_method), HQGeneratorMethod)` might make them slightly easier to understand that way

mjriley commented on corehq/apps/settings/tests/test_views.py:
```diff
@@ -105,29 +107,39 @@ def _create_view_for_user(self, user):
+        self._create_view_for_user(user)
+
+        generator_method = registry.get_method('generator')
+        self.assertEqual(HQGeneratorMethod, type(generator_method))
```
I wonder if there's a more intuitive way to test this. I'm guessing our previous code required setting up a view to verify the 2FA process, but it seems the new API has changed that where we'd only need to call something like `setup_2fa_registry()` before making these verifications. Would that make sense to configure the setup process that way?

gherceg replied:
This was removed in 4ccbb9f9c74ee16ae5179cd22299b53e1dbdcabf. I think the tests now operate similar to how you've specified, asserting that after calling `setup_2fa_registry`, the values are as expected.

mjriley commented on corehq/apps/settings/views.py:
```diff
@@ -403,26 +401,31 @@ class TwoFactorSetupView(BaseMyAccountView, SetupView):
+        # but SetupView.get_form_list hard codes the default validation form for 'sms'
+        # and 'call' methods.
+        # https://github.com/jazzband/django-two-factor-auth/blob/1.15.5/two_factor/views/core.py#L510-L511
+        form_list = super(TwoFactorSetupView, self).get_form_list()
```
nit: use `super().method()` rather than `super(class, self).method()` ([super docs](https://docs.python.org/3/library/functions.html#super))

gherceg replied:
d94b897253b993dbba6cd400627413b6f79daf6b

mjriley commented on corehq/apps/settings/views.py:
```diff
@@ -538,17 +541,11 @@ def dispatch(self, request, *args, **kwargs):
     def get(self, request, *args, **kwargs):
         default_device(request.user).delete()
+        # django-two-factor-auth caches the default device on the user so clear that after deleting
+        from two_factor.utils import USER_DEFAULT_DEVICE_ATTR_NAME
+        delattr(request.user, USER_DEFAULT_DEVICE_ATTR_NAME)
```
Should we have a test verifying this behavior?

gherceg replied:
Ended up understanding this a bit better, and decided to avoid the `default_device` method two_factor provides since it has a side effect of caching the device. Instead I now manually fetch the device and delete it and added tests for this behavior in fd5332d7b42a46a7f5f99363944789176400fb63 (the commit messsage also has some context).

mjriley commented on corehq/apps/hqwebapp/views.py:
```diff
@@ -478,12 +478,17 @@ def iframe_sso_login_pending(request):
-        ('auth', EmailAuthenticationForm),
-        ('token', HQAuthenticationTokenForm),
-        ('backup', HQBackupTokenForm),
+        (LoginView.AUTH_STEP, EmailAuthenticationForm),
+        (LoginView.TOKEN_STEP, HQAuthenticationTokenForm),
+        (LoginView.BACKUP_STEP, HQBackupTokenForm),
```
Appreciate the use of constants here

mjriley commented on corehq/apps/hqwebapp/apps.py:
```diff
@@ -7,3 +8,35 @@ class HqWebAppConfig(AppConfig):
+
+
+def register_two_factor_methods():
+    from two_factor.plugins.registry import registry
```
Is this an import that can be moved to the top of the module? I'd like to conform to [PEP8](https://peps.python.org/pep-0008/#imports) when possible

gherceg replied:
Thanks yeah I moved it up in 2543f363fb65820a1cf26a038e5b90e73aae1c56. When testing locally I got a `AppRegistryNotReady` error when attempting to import the methods in `two_factor_methods.py` at the top of the module because that file imports django_otp, which is not ready when the app ready handler is called here (not quite sure why that is as `django_otp` is included prior to this app, but didn't feel worth digging into more as local imports in an app ready handler are common).

mjriley commented on corehq/apps/hqwebapp/apps.py:
```diff
@@ -7,3 +8,35 @@ class HqWebAppConfig(AppConfig):
+        return
+
+    if getattr(settings, 'TWO_FACTOR_CALL_GATEWAY', None):
+        registry.unregister('call')
```
Just confirming that calling `unregister` on a method that wasn't registered originally won't be an issue?

gherceg replied:
Yeah here's the [implementation](https://github.com/jazzband/django-two-factor-auth/blob/1.15.5/two_factor/plugins/registry.py#L86-L87) in django-two-factor-auth for v1.15.5. It just filters the list of methods to exlude any matching the key passed in, in this case 'call'.

mjriley commented on corehq/apps/hqwebapp/tests/test_apps.py:
```diff
@@ -0,0 +1,46 @@
+    TWO_FACTOR_SMS_GATEWAY='corehq.apps.hqwebapp.two_factor_gateways.Gateway',
+    TWO_FACTOR_CALL_GATEWAY='corehq.apps.hqwebapp.two_factor_gateways.Gateway',
+)
+class RegisterCustom2FAMethodsTests(SimpleTestCase):
```
Thanks for this test suite!

mjriley reviewed (approved):
Looks good!

gherceg committed:
Use as_action where it is possible to have TOTPDevice as other

Initially it is only possible to have phone devices in the
list of other_devices as part of the page context. However,
when a user wants to use their backup phone instead, that
triggers a POST on the accounts/login endpoint which will
generate new context_data, with the TOTPDevice included in
the list of other_devices. This causes an AssertionError
when calling device_action on each device since this is a
phone specific method. Using as_action instead should avoid
the AssertionError.

gherceg committed:
Move two_factor registry import to top to conform with PEP8

two_factor_methods.py needs to be imported locally, otherwise
that file attempts to import django_otp which for some reason
is not ready, despite being defined before hqwebapp.

gherceg committed:
Add conditional to only show other devices if looking at default device

gherceg committed:
Address minor feedback

- setdefault is misleading, just set the key/value on the dict
- use constants where appropriate
- simplify super call

gherceg committed:
Replace default_devices with manual fetch of default device

The default_devices method caches the device object dynamically
on the user object passed in. This side effect of what should be
a simple getter requires us to clean up the cached attribute in this
context since we just want to delete the default device. Therefore,
lets just not use it.

gherceg commented:
I'm picking this back up and not too content with the commit history of this PR. There are some more changes incoming as well, but I'm tempted to rebase this work. @mjriley and @millerdev given you have already reviewed this work, would you prefer I create a new PR if I decide to rework the commit history, or are you ok with me rebasing this PR directly?

gherceg committed:
Reference latest commit of django-two-factor-auth off of 1.16.0

The change will be part of the next release of django-two-factor-auth,
so once the next release is made we can switch back to referencing
releases directly.

This resolves an issue with the phone call method due to invalid XML in
the token.xml template file used by Twilio when providing the code.

gherceg committed:
Override has_token_step to return False if valid backup token has been used

gherceg committed:
Update caller of get_available_phone_methods

This was updated in v1.16.0 of django-two-factor-auth to return
MethodBase objects instead of a tuple of (code, display_text)

gherceg committed:
Add 2FA throttle factors to settings.py

Values are used in django-two-factor-auth and django-otp throttling
logic.

gherceg commented:
QA passed but I uncovered one slight change that I missed previously. In 1.15.4, django-two-factor-auth updated the throttling constants used for rate limiting 2FA login attempts. I've explicitly set these in our settings.py file to ensure a constant of 1, which is what was used previously, is still used to preserve behavior after the upgrade.

gherceg committed:
Remove TOTP Secret from 2FA setup template

If we decide to add the TOTP Secret to this view, it should be done in a
separate PR unrelated to the django-two-factor-auth upgrade.

gherceg commented:
I confirmed that the 2FA setup page displays and functions as expected, and that the throttling timing also works as expected on staging. Going to merge!
